Limit number of transactions passed outside of TransactionController

whymarrh opened the pull request:
Refs #8377
Refs #8572
Refs #8991

This change limits the number of transactions (`txMeta`s) that are passed outside of the `TransactionController`, resulting in shorter serialization and deserialization times when state is moved between the background and UI contexts.

### `TransactionController#_updateMemstore`

The `currentNetworkTxList` state of the `TransactionController` is used externally (i.e. outside of the controller) as the canonical source for the full transaction history. Prior to this change, the method would iterate the full transaction history and possibly return all of it.

This change limits it to `MAX_MEMSTORE_TX_LIST_SIZE` to make sure that:

1. Calls to `_updateMemstore` are fast(er)
2. Passing `currentNetworkTxList` around is fast(er)

(Shown in #8377, `_updateMemstore`, is called _frequently_ when a transaction is pending.)

The list is iterated backwards because it is possible that new transactions are at the end of the list.<sup>[\[1\]][1]</sup>

### Results

In profiles before this change, with ~3k transactions locally, `PortDuplexStream._onMessage` took up to ~4.5s to complete when the set of transactions is included.<sup>[\[2\]][2]</sup>

In profiles after this change, `PortDuplexStream._onMessage` took ~90ms to complete.<sup>[\[3\]][3]</sup>

<details>
<summary>Before vs. after profile screenshots</summary>
<img width="1526" alt="" src="https://user-images.githubusercontent.com/1623628/87613203-36f51d80-c6e7-11ea-89bc-11a1cc2f3b1e.png">
<img width="1526" alt="" src="https://user-images.githubusercontent.com/1623628/87613215-3bb9d180-c6e7-11ea-8d85-aff3acbd0374.png">
</details>

  [1]:https://github.com/MetaMask/metamask-extension/blob/5a3ae85b728096cb45c8cc6822249eed5555ee25/app/scripts/controllers/transactions/tx-state-manager.js#L172-L174
  [2]:https://user-images.githubusercontent.com/1623628/87613203-36f51d80-c6e7-11ea-89bc-11a1cc2f3b1e.png
  [3]:https://user-images.githubusercontent.com/1623628/87613215-3bb9d180-c6e7-11ea-8d85-aff3acbd0374.png
  [8337]:https://github.com/MetaMask/metamask-extension/issues/8377
  [8572]:https://github.com/MetaMask/metamask-extension/issues/8572
  [8991]:https://github.com/MetaMask/metamask-extension/issues/8991

whymarrh committed:
Limit number of transactions passed outside of TransactionController

Refs #8572
Refs #8991

This change limits the number of transactions (`txMeta`s) that are passed
outside of the `TransactionController`, resulting in shorter serialization and
deserialization times when state is moved between the background and UI
contexts.

`TransactionController#_updateMemstore`
---------------------------------------

The `currentNetworkTxList` state of the `TransactionController` is used
externally (i.e. outside of the controller) as the canonical source for
the full transaction history. Prior to this change, the method would iterate
the full transaction history and possibly return all of it.

This change limits it to `MAX_MEMSTORE_TX_LIST_SIZE` to make sure that:

1. Calls to `_updateMemstore` are fast(er)
2. Passing `currentNetworkTxList` around is fast(er)

(Shown in #8377, `_updateMemstore`, is called _frequently_ when a transaction
is pending.)

The list is iterated backwards because it is possible that new transactions are
at the end of the list. [1]

Results
-------

In profiles before this change, with ~3k transactions locally,
`PortDuplexStream._onMessage` took up to ~4.5s to complete when the set of
transactions is included. [2]

In profiles after this change, `PortDuplexStream._onMessage` took ~90ms to
complete. [3]

Before vs. after profile screenshots:

![Profile 1][2]
![Profile 2][3]

  [1]:https://github.com/MetaMask/metamask-extension/blob/5a3ae85b728096cb45c8cc6822249eed5555ee25/app/scripts/controllers/transactions/tx-state-manager.js#L172-L174
  [2]:https://user-images.githubusercontent.com/1623628/87613203-36f51d80-c6e7-11ea-89bc-11a1cc2f3b1e.png
  [3]:https://user-images.githubusercontent.com/1623628/87613215-3bb9d180-c6e7-11ea-8d85-aff3acbd0374.png
  [8337]:https://github.com/MetaMask/metamask-extension/issues/8377
  [8572]:https://github.com/MetaMask/metamask-extension/issues/8572
  [8991]:https://github.com/MetaMask/metamask-extension/issues/8991

brad-decker reviewed (dismissed):
LGTM

Gudahtt reviewed (commented):
This seems kinda dangerous to me. Our UI assumes that all of the transactions in a 'group' (i.e. with the same nonce) are present, and it would display things incorrectly otherwise. e.g. a cancellation might be displayed as sending ether to yourself, or a transaction might show up as 'dropped' if an earlier transaction with the same nonce completed, but wasn't included in this set. Splitting up transaction groups would be nice to avoid.

Maybe instead of counting transactions, we could count `nonce`'s backwards? and keep the first X nonces encountered.

whymarrh commented:
That sounds feasible, I'll take a swing at that.

whymarrh commented:
I've updated the logic here to look for N unique nonces (and added tests!). This still iterates the whole tx list but should account for how the set of transactions are used by the UI.

rekmarks commented on app/scripts/controllers/transactions/tx-state-manager.js:
```diff
@@ -57,12 +57,39 @@ export default class TransactionStateManager extends EventEmitter {
     const fullTxList = this.getFullTxList()
-    return fullTxList.filter((txMeta) => txMeta.metamaskNetworkId === network)
+
+    const nonces = new Set()
```
You love to see a solid use of `Set`s!

rekmarks commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -37,6 +37,7 @@ import { hexToBn, bnToHex, BnMultiplyByFraction } from '../../lib/util'
 import { TRANSACTION_NO_CONTRACT_ERROR_KEY } from '../../../../ui/app/helpers/constants/error-keys'
 
 const SIMPLE_GAS_COST = '0x5208' // Hex for 21000, cost of a simple send.
+const MAX_MEMSTORE_TX_LIST_SIZE = 100 // Number of transactions (by unique nonces) to keep in memory
```
Is this number too conservative for our power users? I don't have anywhere near 3000 transactions, so I don't know what they might want/need. Presumably our UI can handle 150-200 without slowing down, if 100 took (IIRC) 90ms.

I am working on fixing the state log explorer, so they should have that by 8.0.5, regardless.

whymarrh replied:
I don't know, I don't think there's _a lot_ of utility in looking at the tx details 100+ txs back, but yeah I'm just guessing. A user can still go to Etherscan/their favourite block explorer to see details past this.

whymarrh replied:
We can also change this in the future since this doesn't change how things are persisted.

rekmarks reviewed (approved):
LGTM!

Gudahtt reviewed (approved):
LGTM!
